### PR TITLE
on-demand, upstream: on-demand cluster discovery

### PR DESCRIPTION
**Pull request by krnowak**
I'm filing an work-in-progress PR for the feature to gather some initial feedback on the code. Right now I'm working on writing tests.

Commit Message:
This PR adds an on-demand cluster discovery feature to Envoy. The process of the discovery is driven through the on-demand filter. The on-demand filter is updated to also use a per-route configuration to enable ODCDS.

Additional Description:
There is a design document for the feature - https://docs.google.com/document/d/1AqhQnrX_7SS6PAkNwoaA2peCtk9htgwSBTRg1gzFPNw/edit#

I haven't yet implemented checking if CDS is using delta, as ODCDS should not be used together with SotW CDS.

Risk Level:
Medium (new feature)

Testing:
Currently writing tests, will also do some manual testing with cluster_header in RouteAction.

Docs Changes:
TODO (should be both API docs and some usage documentation in on-demand filter, I think).

Release Notes:
New feature: on_demand: On demand discovery of the clusters.

Platform Specific Features:
N/A

[Optional Runtime guard:]

[Optional Fixes #Issue]
Partially addresses #2500 (the clusters part).

[Optional Deprecated:]

[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/extensions/filters/http/on_demand/v3/on_demand.proto [hunk 1]
```diff
@@ -10,11 +12,19 @@ option java_outer_classname = "OnDemandProto";
...
+// [#protodoc-title: On Demand Discovery]
+// On Demand Discovery :ref:`configuration overview <config_http_filters_on_demand>`.
 // [#extension: envoy.filters.http.on_demand]
 
+// On Demand Discovery filter config.
 message OnDemand {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.config.filter.http.on_demand.v2.OnDemand";
+
+  config.core.v3.ConfigSource odcds_config = 1;
```
Comment (mattklein123) [inline api/envoy/extensions/filters/http/on_demand/v3/on_demand.proto:24]: docs please, including what happens if not configured. Same below.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks, at a very high level this looks like it's directionally correct. Please let me know when this is ready for real review (comments, tests, etc.).

/wait

File: source/common/upstream/cluster_manager_impl.cc [hunk 2]
```diff
@@ -1018,6 +1019,172 @@ ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks&
...
+}
+
+ClusterDiscoveryCallbackHandlePtr
+ClusterManagerImpl::requestOnDemandClusterDiscovery(CdsApiWeakPtr weak_cds, const std::string& name,
+                                                    ClusterDiscoveryCallbackWeakPtr weak_callback) {
+  ThreadLocalClusterManagerImpl& cluster_manager = *tls_;
+
+  cluster_manager.cdm_.ensureCallbacksAreInstalled();
+
+  auto [handle, discovery_in_progress] = cluster_manager.cdm_.addCallback(name, weak_callback);
```
Comment (dmitri-d) [inline source/common/upstream/cluster_manager_impl.cc:1127]: This looks like it will miss discovery requests for `name` initiated in other threads, is this intentional? Could you add a comment then?

File: source/extensions/filters/http/on_demand/on_demand_update.cc [hunk 3]
```diff
@@ -4,16 +4,41 @@
...
 namespace HttpFilters {
 namespace OnDemand {
 
+OnDemandFilterConfig::OnDemandFilterConfig(
+    const envoy::extensions::filters::http::on_demand::v3::OnDemand& proto_config,
+    Upstream::ClusterManager& cm, Stats::Scope& scope,
+    ProtobufMessage::ValidationVisitor& validation_visitor)
+    : cm_(cm) {
+  if (proto_config.has_odcds_config()) {
+    odcds_ =
```
Comment (dmitri-d) [inline source/extensions/filters/http/on_demand/on_demand_update.cc:22]: This creates an instance of CDS api/subscription in a worker thread, which is different from the rest of xDS that runs on the main thread. I think this is an issue -- there are concurrency (definitely with ADS configured) and synchronisation issues here. I probably missed it -- where is the CDS subscription started?

**Comment by dmitri-d** [comment]
Would it be possible for you to write a brief explanation re: how you plan to handle configuration and starting of on-demand CDS subscriptions? Did you have a chance to read https://blog.envoyproxy.io/envoy-threading-model-a8d44b922310?

File: source/extensions/filters/http/on_demand/on_demand_update.cc (hunk 3 above)
Comment (krnowak) [reply source/extensions/filters/http/on_demand/on_demand_update.cc:22]: This is a good question. I'm still in progress of writing the tests, so this will come out soon. But yeah, looks like creating CDS in worker thread calls into cluster manager directly, so looks like I either would need to make it thread safe, or postpone the creation of cds until I'm in main thread.

Basically I didn't want to start the subscription, because it will result in periodically requesting the management server, which is not what I wanted. I'll have a look what VHDS is doing here.

File: source/common/upstream/cluster_manager_impl.cc (hunk 2 above)
Comment (krnowak) [reply source/common/upstream/cluster_manager_impl.cc:1127]: Will add more comments, sure.

Note that `addCallback` will add a callback anyway. But if the discovery for `name` was already initiated in this thread, then we don't need to do anything else, just add the callback and wait for it to be fired. If the discovery for `name` was initiated in some other worker thread, then we have no way of knowing that in this worker thread, so we try to initiate the discovery anyway. Whether the discovery will actually be initiated is something that main thread will figure out (see the code where we check `pending_cluster_creations_` in main thread).

File: source/extensions/filters/http/on_demand/on_demand_update.cc (hunk 3 above)
Comment (krnowak) [reply source/extensions/filters/http/on_demand/on_demand_update.cc:22]: We did a bit of tracking where this is called from and I have an impression that both `createRouteSpecificFilterConfigTyped` and `createFilterFactoryFromProtoTyped` are called from the main thread. And inside those functions we create the CDS, so it ought to be fine. Maybe the factories that those functions return are called from worker threads, aren't they?

Comment (dmitri-d) [reply source/extensions/filters/http/on_demand/on_demand_update.cc:22]: > Basically I didn't want to start the subscription, because it will result in periodically requesting the management server, which is not what I wanted. I'll have a look what VHDS is doing here.

A subscription is basically type-specific (i.e. CDS, RDS, VHDS, etc) grpc connection to a management server; in case of ADS all subscriptions are multiplexed over a single grpc connection. The connection needs to be initiated (subscription started) in order to be able to both send and receive xDS updates.

Note that VHDS establishes a separate subscription in addition to the base RDS one and calculates a common state (based on RDS state and VHDS updates). I encourage you to take a closer look at various xDS implementations, as well as muxes. Please also take a look at the docs I linked to earlier.

File: api/envoy/extensions/filters/http/on_demand/v3/on_demand.proto (hunk 1 above)
Comment (markdroth) [inline api/envoy/extensions/filters/http/on_demand/v3/on_demand.proto:24]: If the only knob here is the presence of a `ConfigSource`, then how will we configure ODCDS when we're using new-style resource names and don't need a `ConfigSource` because we're using authority-based routing?

Comment (htuch) [reply api/envoy/extensions/filters/http/on_demand/v3/on_demand.proto:24]: I think the presence of an OCDS `PerRouteConfig` would indicate we needed to treat the name as on-demand. The meaning of `self` becomes interesting as it might mean two different things for the filter-level config (received from LDS) and the `PerRouteConfig` (from RDS).

**Comment by krnowak** [comment]
Hi, updated the PR. I still have a hitch with the integration test failing. Not sure if this is something wrong with my changes in cluster manager or the test itself is wrong. I have asked on slack on how to enable logs during running the integration tests - this hopefully would help me with debugging the issue.

To answer some questions:

@dmitri-d: I wrote a separate class instead of using CdsApi{,Impl} - it shares most of the functionality with CdsApiImpl, but it starts the subscription on first use instead of up-front.

@markdroth: Unfortunately I'm not very well oriented in matters of new-style cluster naming and authority-based routing. I was pointed to the document on the last community meeting, so I'll try to get through it soon.

**Comment by krnowak** [comment]
Got a bit further (it was a `new-cluster` vs `new_cluster` mixup). Now the test seems to crash on teardown:

```
 Caught Segmentation fault, suspect faulting address 0x0
 Backtrace (use tools/stack_decode.py to get line numbers):
 Envoy version: 0/1.18.0-dev/test/DEBUG/BoringSSL
 #0: Envoy::SignalAction::sigHandler() [0x4bc4dac]
 #1: __restore_rt [0x7f0e7a1f4980]
 #2: Envoy::Config::NewGrpcMuxImpl::trySendDiscoveryRequests() [0x44e7cbf]
 #3: Envoy::Config::NewGrpcMuxImpl::updateWatch() [0x44ea82b]
 #4: Envoy::Config::NewGrpcMuxImpl::removeWatch() [0x44eaed4]
 #5: Envoy::Config::NewGrpcMuxImpl::WatchImpl::remove() [0x44f859d]
 #6: Envoy::Config::NewGrpcMuxImpl::WatchImpl::~WatchImpl() [0x44f848a]
 #7: Envoy::Config::NewGrpcMuxImpl::WatchImpl::~WatchImpl() [0x44f850c]
 …
 #10: std::__1::unique_ptr<>::~unique_ptr() [0x2e18149]
 #11: Envoy::Config::GrpcSubscriptionImpl::~GrpcSubscriptionImpl() [0x44c1a55]
 #12: Envoy::Config::GrpcSubscriptionImpl::~GrpcSubscriptionImpl() [0x44c1abc]
 …
 #15: std::__1::unique_ptr<>::~unique_ptr() [0x2dfc229]
 #16: Envoy::Upstream::OdCdsApiImpl::~OdCdsApiImpl() [0x404f8e9]
 #17: Envoy::Upstream::OdCdsApiImpl::~OdCdsApiImpl() [0x404f93c]
 …
 #20: std::__1::__shared_weak_count::__release_shared() [0x60ccd8e]
 #21: Envoy::Extensions::HttpFilters::OnDemand::OnDemandFilterConfig::~OnDemandFilterConfig() [0x4035ed5]
 …
 #34: absl::node_hash_map<>::~node_hash_map() [0x4596fb8]
 #35: Envoy::Router::PerFilterConfigs::~PerFilterConfigs() [0x4592ed5]
 #36: Envoy::Router::VirtualHostImpl::~VirtualHostImpl() [0x45993db]
 #37: Envoy::Router::VirtualHostImpl::~VirtualHostImpl() [0x459949c]
 …
 #40: std::__1::__shared_weak_count::__release_shared() [0x60ccd8e]
 #41: Envoy::Router::RouteMatcher::~RouteMatcher() [0x41bb7f6]
 …
 #44: std::__1::unique_ptr<>::~unique_ptr() [0x41bb6b9]
 #45: Envoy::Router::ConfigImpl::~ConfigImpl() [0x41bb68d]
 #46: Envoy::Router::ConfigImpl::~ConfigImpl() [0x45995ac]
 …
 #49: std::__1::__shared_weak_count::__release_shared() [0x60ccd8e]
 #50: Envoy::Router::StaticRouteConfigProviderImpl::~StaticRouteConfigProviderImpl() [0x41c8cc6]
 #51: Envoy::Router::StaticRouteConfigProviderImpl::~StaticRouteConfigProviderImpl() [0x41c8d3c]
 …
 #54: std::__1::__shared_weak_count::__release_shared() [0x60ccd8e]
 #55: Envoy::Extensions::NetworkFilters::HttpConnectionManager::HttpConnectionManagerConfig::~HttpConnectionManagerConfig() [0x4077847]
 #56: std::__1::__shared_ptr_emplace<>::__on_zero_shared() [0x40871d6]
 #57: std::__1::__shared_weak_count::__release_shared() [0x60ccd8e]
 #58: Envoy::Extensions::NetworkFilters::HttpConnectionManager::HttpConnectionManagerFilterConfigFactory::createFilterFactoryFromProtoTyped()::$_6::~createFilterFactoryFromProtoTyped() [0x4060516]
 …
```

I tried deferring the destruction to main thread (just to be sure), but it crashes at the same place too. Will investigate.

**Comment by krnowak** [comment]
Fixed the crash (it was caused by bad member ordering in OdCdsApiImpl). Should be ready for a review.

**Comment by krnowak** [comment]
This PR should be ready for review. The only issue CI raised was insufficient coverage, so I added more tests to fix this and I rebased the branch to current main to fix a conflict. Thanks.

**Comment by mattklein123** [comment]
@krnowak from this point forward please never force push. It breaks reviews. Thank you.

File: include/envoy/upstream/cluster_manager.h [hunk 4]
```diff
@@ -309,6 +328,32 @@ class ClusterManager {
...
+   *
+   * This function is thread-safe.
+   *
+   * @param weak_odcds is a weak pointer to the ODCDS that will be used for discovery.
+   * @param name is the name of the cluster to be discovered.
+   * @param callback will be called when the discovery is finished.
+   * @return the discovery process handle
+   */
+  virtual ClusterDiscoveryCallbackHandlePtr
+  requestOnDemandClusterDiscovery(OdCdsApiWeakPtr weak_odcds, const std::string& name,
```
Comment (mattklein123) [inline include/envoy/upstream/cluster_manager.h:347]: It's odd to pass a weak_ptr here. I would pass a shared_ptr since the ocds should exist at this point and allow the callee to convert to weak_ptr as needed.

File: include/envoy/upstream/cluster_manager.h [hunk 5]
```diff
@@ -309,6 +328,32 @@ class ClusterManager {
...
+                                  ClusterDiscoveryCallbackWeakPtr callback) PURE;
+
+  /**
+   * Notifies the threads that were waiting for discovery of a cluster with a passed name. Usually
+   * used to unblock the requests in case of discovery failure.
+   *
+   * @param name is the cluster's name.
+   */
+  virtual void notifyOnDemandCluster(const std::string& name, bool cluster_exists) PURE;
```
Comment (mattklein123) [inline include/envoy/upstream/cluster_manager.h:356]: It seems odd to me that this is a public method? Why is this needed? Shouldn't this be an implementation detail of the internal process that does on demand discovery?

File: source/common/upstream/cds_api_helper.h [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+#include "envoy/config/subscription.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "common/common/logger.h"
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Upstream {
+
+class CdsApiHelper : Logger::Loggable<Logger::Id::upstream> {
```
Comment (mattklein123) [inline source/common/upstream/cds_api_helper.h:15]: Please add some class comments on what this class is for to aid review/readability.

File: source/common/upstream/cds_api_impl.cc [hunk 7]
```diff
@@ -71,56 +59,8 @@ void CdsApiImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef>& r
 void CdsApiImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef>& added_resources,
                                 const Protobuf::RepeatedPtrField<std::string>& removed_resources,
                                 const std::string& system_version_info) {
-  Config::ScopedResume maybe_resume_eds;
```
Comment (mattklein123) [inline source/common/upstream/cds_api_impl.cc:74]: To aid review, can you split this pure refactor out into a separate PR? Basically just make the helper class and move the code, and then we can get back to this PR?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks for working on this. Very exciting. Flushing out a first round of comments. 

/wait

**Comment by WeavingGao** [comment]
I am looking forward to this feature, if there is any further progress？

**Comment by krnowak** [comment]
> I am looking forward to this feature, is there any follow-up progress?

Yes, I'm working on it.

File: include/envoy/upstream/cluster_manager.h (hunk 4 above)
Comment (krnowak) [reply include/envoy/upstream/cluster_manager.h:347]: Yeah, will pass a shared_ptr for odcds here. Now looking at the code, there are probably some corner cases with the weak_ptr, so change to shared_ptr should make it simpler and less cornery. But I'll keep the weak_ptr for callback.

File: include/envoy/upstream/cluster_manager.h (hunk 5 above)
Comment (krnowak) [reply include/envoy/upstream/cluster_manager.h:356]: I didn't have a clear idea how to signal a failed discovery from ODCDS (not only the case where `onConfigUpdateFailed` is called, but also when requesting a cluster and it's repeatedly absent in response). ODCDS takes `ClusterManager&` as a parameter, not `ClusterManagerImpl`, so I needed to make `notifyOnDemandCluster` a part of `ClusterManager` API. This notification is used to effectively unpause the paused request in the on-demand extension.

I suppose that instead of this `notifyOnDemandCluster` I could install a timer for each discovery process that invokes the callbacks in each thread. That would simplify the ODCDS code and make the `notifyOnDemandCluster` function unnecessary. I only wonder what the timeout should be? Something specified in the on-demand configuration? Or a hardcoded value?

File: source/common/upstream/cds_api_helper.h (hunk 6 above)
Comment (krnowak) [reply source/common/upstream/cds_api_helper.h:15]: Addressed in https://github.com/envoyproxy/envoy/pull/15806.

File: source/common/upstream/cds_api_impl.cc (hunk 7 above)
Comment (krnowak) [reply source/common/upstream/cds_api_impl.cc:74]: Please see https://github.com/envoyproxy/envoy/pull/15806.

File: include/envoy/upstream/cluster_manager.h (hunk 5 above)
Comment (krnowak) [reply include/envoy/upstream/cluster_manager.h:356]: I made a change in a commit here: https://github.com/kinvolk/envoy/commit/b94a489c9cbcfedde9a36dd73479ff392f763adb, but I haven't yet pushed it - building it locally and making sure that it works.

Comment (mattklein123) [reply include/envoy/upstream/cluster_manager.h:356]: OK I can take a look at your updated change when you push it. In general I think a callback on failure is fine, but the way I would implement it is to pass a callback object that is also notified on failure. At that point it's an internal implementation detail and you shouldn't need this method.

**Comment by krnowak** [comment]
Now that #15806 is merged, do you expect me to merge the main branch into the PR branch and fix the conflicts?

**Comment by mattklein123** [comment]
> Now that #15806 is merged, do you expect me to merge the main branch into the PR branch and fix the conflicts?

Yes. I don't know how else we would proceed. :)

**Comment by krnowak** [comment]
> > Now that #15806 is merged, do you expect me to merge the main branch into the PR branch and fix the conflicts?
> 
> Yes. I don't know how else we would proceed. :)

Heh, right. Projects I was working on so far allowed rebasing, so I just wanted to be sure. :)

Anyway, I updated the PR. I think I have addressed most (or all) of the initial review issues.

File: source/common/upstream/odcds_api_impl.cc [hunk 8]
```diff
@@ -0,0 +1,85 @@
+#include "common/upstream/odcds_api_impl.h"
```
Comment (htuch) [inline source/common/upstream/odcds_api_impl.cc:1]: Sorry, I'm a bit late to the party here, but why do we have a different CDS implementation for on-demand vs. regular CDS? In all the other on-demand models (VHDS, SRDS), we reuse the existing machinery and just making it on-demand aware. Here we have diverged CDS implementation. This will probably have implications down the track when we make bug fixes for regular delta CDS but miss out on on OCDS and vice versa.

I would like to see better awareness of on-demand as a horizontal in Envoy that spans across subsystems, rather than doing what is expedient or local to CDS for OCDS here. 

CC @mattklein123 @adisuissa @dmitri-d

File: source/common/upstream/cluster_manager_impl.cc [hunk 9]
```diff
@@ -1027,6 +1028,185 @@ ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks&
...
+  if (discovery_in_progress) {
+    // We have already started a discovery process for a cluster with
+    // this name, so nothing more left to do here.
+    return std::move(handle);
+  }
+  dispatcher_.post([this, odcds = std::move(odcds), weak_callback, name,
```
Comment (htuch) [inline source/common/upstream/cluster_manager_impl.cc:1150]: As per my other comment, I think this is tricky cross-thread lifetime related code for on-demand that has _already_ been addressed in both VHDS and SRDS. Re-inventing these patterns here is problematic, we are building guaranteed technical debt.

Sorry to be a downer, I appreciate the work being done here, but I know from experience how hard it was to get things like on-demand VHDS right, we need to have common patterns for deferral and resumption across the on-demand xDS.

CC @mattklein123 @adisuissa @dmitri-d

File: source/common/upstream/odcds_api_impl.cc (hunk 8 above)
Comment (mattklein123) [reply source/common/upstream/odcds_api_impl.cc:1]: > CC @mattklein123 @adisuissa @dmitri-d

As I said offline, I haven't actually reviewed the code in detail yet, just the interfaces. I will review in more detail today. My expectation is most/all of the code is shared. I'm not sure it is possible to implement on-demand in main CDS due to the SoTW wildcard semantics, but we can see (at a certain point though it may just mean moving the same code split to a different location).

File: include/envoy/upstream/cluster_manager.h [hunk 10]
```diff
@@ -339,6 +393,21 @@ class CdsApi {
...
+public:
+  virtual ~OdCdsApi() = default;
+
+  /**
+   * File an on-demand request for a cluster.
+   */
+  virtual void updateOnDemand(const std::string& cluster_name) PURE;
+};
+
+using OdCdsApiPtr = std::unique_ptr<OdCdsApi>;
```
Comment (mattklein123) [inline include/envoy/upstream/cluster_manager.h:409]: Somewhat related to @htuch comments, but why is this public interface needed? Conceptually, something asks the cluster manager to fetch a cluster, and gets a callback when complete or failed. I don't think any of this should be exposed "externally" and should be an internal implementation detail.

File: source/common/upstream/cluster_manager_impl.cc (hunk 9 above)
Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc:1150]: Yes I agree if we can share this out that would be good.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Hi, thanks for the updates. This PR is still too large. For the next set of changes can we:
1) Split out only the cluster manager changes and tests around those changes.
2) Save the on-demand filter changes for the next PR.

I think if we do (1) we can more easily evaluate @htuch concerns as well as my high level interface concerns. Thank you!

/wait

File: api/envoy/extensions/filters/http/on_demand/v3/on_demand.proto [hunk 11]
```diff
@@ -10,11 +12,31 @@ option java_outer_classname = "OnDemandProto";
...
+  // service. If not specified, the on-demand cluster discovery will
+  // be disabled. When it's specified, the filter will pause a request
+  // to an unknown cluster and will begin a cluster discovery
+  // process. When the discovery is finished (successfully or not),
+  // the request will be resumed for further processing.
+  config.core.v3.ConfigSource odcds_config = 1;
+}
+
+// Per-route configuration for On Demand Discovery.
+message PerRouteConfig {
```
[the remark this answers is not in the document]
Comment (markdroth) [reply api/envoy/extensions/filters/http/on_demand/v3/on_demand.proto:34]: We need both the ability to disable this on a per-route basis and the ability to handle xdstp names for which no `ConfigSource` needs to be specified.  In https://github.com/envoyproxy/envoy/pull/15523/files/b742a88722c205b9fa1d0b694ff010c3d994e7b2#r596505958, @htuch proposed that we use an empty `PerRouteConfig` message for xdstp names; here, you are proposing that we use the same thing to indicate disabling the feature.  I don't think this can mean both things.

**Comment by markdroth** [comment]
I'd like a bit more info on exactly which problem we're trying to solve here.  As I mentioned in [this comment thread in the design doc](https://docs.google.com/document/d/1AqhQnrX_7SS6PAkNwoaA2peCtk9htgwSBTRg1gzFPNw/edit?disco=AAAALssbA0M), I think there are two closely related but distinct problems in this space, and it's not clear to me what our strategy is for each of the two.  I'm concerned that we may be conflating them here in a way that may limit flexibility.

File: api/envoy/extensions/filters/http/on_demand/v3/on_demand.proto (hunk 11 above)
[the remark this answers is not in the document]
Comment (htuch) [reply api/envoy/extensions/filters/http/on_demand/v3/on_demand.proto:34]: Technically we could have an empty message with an empty `ConfigSource`, but I think this gets confusing. Probably an explicit disable flag would be the way to go.

File: source/common/upstream/odcds_api_impl.cc (hunk 8 above)
Comment (krnowak) [reply source/common/upstream/odcds_api_impl.cc:1]: > Sorry, I'm a bit late to the party here, but why do we have a different CDS implementation for on-demand vs. regular CDS?

There is not much of the code difference really - most of the code is shared through the CdsApiHelper. The differences between CDS and ODCDS are mostly about when to start subscription (for CDS is sometime at Envoy startup, for ODCDS is on the first request). I suppose I could make CDS on-demand aware, but I'm afraid that would result in awkward code inside it.

> In all the other on-demand models (VHDS, SRDS), we reuse the existing machinery and just making it on-demand aware. Here we have diverged CDS implementation. This will probably have implications down the track when we make bug fixes for regular delta CDS but miss out on on OCDS and vice versa.

I tried to move most of the code into the helper class so such bug fixes would most certainly land in a helper class, not in CDS/ODCDS itself.

> 
> I would like to see better awareness of on-demand as a horizontal in Envoy that spans across subsystems, rather than doing what is expedient or local to CDS for OCDS here.
> 
> CC @mattklein123 @adisuissa @dmitri-d

File: source/common/upstream/cluster_manager_impl.cc (hunk 9 above)
Comment (krnowak) [reply source/common/upstream/cluster_manager_impl.cc:1150]: I had a look and what VHDS is doing and I can see the differences:

- I'm "deduplicating" the requests, so 2 worker threads asking for the cluster "foo" will result in one request instead of two.
- I'm installing a timeout on the discovery, so the callbacks list won't grow indefinitely in case of cluster "foo" nonexistence.

I guess I can simplify the code, but maybe let's discuss it in the other PR, where I split out the CM changes.

File: include/envoy/upstream/cluster_manager.h (hunk 10 above)
Comment (krnowak) [reply include/envoy/upstream/cluster_manager.h:409]: Since the config source for the on-demand discovery is provided in the per-route config, which makes it a separate object than CM has from `cds_config`, I need to pass a "CDS" object to CM. It needs to be a public interface, because it's being used in `ClusterManager` interface. I had two options here: either extend the `CdsApi` interface with the `updateOnDemand` function and pass that to CM or create a new interface (`OdCdsApi`). I went with the latter, because I found the former to be awkward. It's because `CdsApi` provides `initialize`, `setInitializedCb` and `versionInfo` and for the on-demand discovery I would use none of it.

**Comment by krnowak** [comment]
Hi,

> Hi, thanks for the updates. This PR is still too large. For the next set of changes can we:
> 
>     1. Split out only the cluster manager changes and tests around those changes.

Please see #15857.

> 
>     2. Save the on-demand filter changes for the next PR.

I suppose I could reuse this PR to add the on-demand filter changes after #15857 settles.

> 
> 
> I think if we do (1) we can more easily evaluate @htuch concerns as well as my high level interface concerns. Thank you!
> 
> /wait

**Comment by krnowak** [comment]
> I'd like a bit more info on exactly which problem we're trying to solve here. As I mentioned in [this comment thread in the design doc](https://docs.google.com/document/d/1AqhQnrX_7SS6PAkNwoaA2peCtk9htgwSBTRg1gzFPNw/edit?disco=AAAALssbA0M), I think there are two closely related but distinct problems in this space, and it's not clear to me what our strategy is for each of the two. I'm concerned that we may be conflating them here in a way that may limit flexibility.

What I'm trying to solve here is to enable (or disable) on-demand cluster discovery on connection manager or virtual host or route level, especially together with cluster_header route action.

What I'm not covering in this PR is the case where I statically specify a cluster name for some route and expect it to be discovered at a time of request handling instead of at Envoy startup.

**Comment by markdroth** [comment]
@krnowak Sorry, I don't quite follow.  If this will not work for routes with static cluster names, then isn't it *only* for the cluster_header route action?  The only three options for any route action are a cluster name, a `WeightedClusters` list (which is basically just a list of static cluster names), and cluster_header.

**Comment by krnowak** [comment]
> @krnowak Sorry, I don't quite follow. If this will not work for routes with static cluster names, then isn't it _only_ for the cluster_header route action? The only three options for any route action are a cluster name, a `WeightedClusters` list (which is basically just a list of static cluster names), and cluster_header.

I'd imagine that "static" cluster names (or weighted clusters) would work if they were provided by RDS or VHDS, for example. But honestly, I haven't checked that yet.

I think the reason I have put a boolean flag or a config source inside the route in my first version of the code was to have an access to them when checking the static clusters [here](https://github.com/envoyproxy/envoy/blob/main/source/common/router/config_impl.cc#L1245) - I could basically disable the check and discover the static cluster at a request time. But now that this setting was moved to the extension config I had no clue how to check if the route has the on-demand discovery enabled from that place.

**Comment by krnowak** [comment]
I'll open a new one with the extension changes after we settle #15857.
